### ISSUE TITLE
Tail / trailer / footer

### ISSUE THREAD

**Issue by dlabz** [kaitai-io/kaitai_struct#912]
TL;DR After not finding an answer in documentation or examples, I decided to waste more of my and some of your time.

Reading the documentation, I was unable to locate a construct to declare a file tail. Such constructs are common in web streams and, more importantly, introducing them in kaitai structs might improve the syntax and performance. 

### example 1: gif files always end with `0x3B`

This is a somewhat trivial example, but I chose it as it's featured in kaitai Quick start. 

From [gif specification](http://giflib.sourceforge.net/gif89.txt):
```27. Trailer.

      a. Description. This block is a single-field block indicating the end of
      the GIF Data Stream.  It contains the fixed value 0x3B.

      b. Required Version.  87a.

      c. Syntax.

      7 6 5 4 3 2 1 0        Field Name                    Type
     +---------------+
  0  |               |       GIF Trailer                   Byte
     +---------------+

      d. Extensions and Scope. This block does not have scope, it terminates
      the GIF Data Stream. This block may not be modified by any extension.

      e. Recommendations. None.
```
from [gif.ksy](https://github.com/kaitai-io/kaitai_struct_formats/blob/master/image/gif.ksy) :

```...
repeat-until: _io.eof or _.block_type == block_type::end_of_file
...
enums:
  block_type:
    0x21: extension
    0x2c: local_image_descriptor
    0x3b: end_of_file
```
### example 2: Autodesk forge pack file

```//TODO: example in ksy```

In this example, tail is used for easier identification of file sections. Last two chunks are `u4` position offsets of types section and toc section respectively, both of which are encoded using `varint`/`vlq_base32` and variable length string. making it impossible to guess the size of each section from information in header.

_The pack file itself is a collection of small binary entries, like vectors, quaternions and matrices, which can't be efficiently transferred over http (transfer overhead) or stored on hdd (block size). To solve this with minimal memory bloat, a generic header is sent, followed by a stream of entries, nothing the type and offset in the type enumerator and toc, which are sent after the body. Tail is added to identify offsets of toc and types tables, making it possible to substream sections._ 

### Footers in popular formats

Extension | Header (Hex) | Footer (Hex)
-- | -- | --
DOC | D0 CF 11 E0 A1 B1 1A E1 | 57 6F 72 64 2E 44 6F 63 75 6D 65 6E 74 2E
XLS | D0 CF 11 E0 A1 B1 1A E1 | FE FF FF FF 00 00 00 00 00 00 00 00 57 00 6F 00 72 00 6B 00 62 00 6F 00 6F 00 6B 00
PPT | D0 CF 11 E0 A1 B1 1A E1 | 50 00 6F 00 77 00 65 00 72 00 50 00 6F 00 69 00 6E 00 74 00 20 00 44 00 6F 00 63 00 75 00 6D 00 65 00 6E 00 74
ZIP | 50 4B 03 04 14 | 50 4B 05 06 00
JPG | FF D8 FF E0 00 10 4A 46 49 46 00 01 01 | D9 (“Better To Use File size Check”)
GIF | 47 49 46 38 39 61 4E 01 53 00 C4 | 21 00 00 3B 00
PDF | 25 50 44 46 2D 31 2E | 25 25 45 4F 46

[src](https://www.datadoctor.biz/data_recovery_programming_book_chapter14-page2.html)

## on the web

When such file is accessed over http, [Range header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range) allows fetching file sections in reverse order, allowing the stream of entries to be parsed on the fly, by indicating negative length, which is read from the end of file. 
```Range: <unit>=<range-start>-
Range: <unit>=<range-start>-<range-end>
Range: <unit>=<range-start>-<range-end>, <range-start>-<range-end>
Range: <unit>=<range-start>-<range-end>, <range-start>-<range-end>, <range-start>-<range-end>
Range: <unit>=-<suffix-length>```

Tail is also used in chunked transfers, by use of [Trailer header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer). 

### Random thoughts 

Using negative offsets is a common way to indicate that length is to be read from the end of, say, string. 

In kaitai, negative offset is more valuable as an indicator of an error in reading sequence and was clear choice not to allow it. 

Still, to count from the end of file / section / chunk can have number of clever uses, but seems to come with a performance penalty, and I wasn't able to find a way to simply fetch the tail of a file, so here we are. 

There are some other uses for a tail in the scope of the most anticipated feature, serialization, as described in the autodesk pack-file section.
